File: frontend/src/types/authSources.ts
import { z } from "zod";

export const AuthSourceSchema = z.object({
  id: z.string(),
  name: z.string(),
  kind: z.enum(["local", "oidc"]),
  enabled: z.boolean(),
  description: z.string(),
  issuer_url: z.string().nullable().optional(),
  client_id: z.string().nullable().optional(),
  has_client_secret: z.boolean(),
  scopes: z.string(),
  claim_username: z.string(),
  claim_groups: z.string(),
  created_at: z.string(),
  updated_at: z.string(),
});

export const UpnSuffixSchema = z.object({
  id: z.string(),
  suffix: z.string(),
  auth_source_id: z.string(),
  auth_source_name: z.string(),
  auth_source_kind: z.string(),
  created_at: z.string(),
  updated_at: z.string(),
});

export const AppIdentityProviderSchema = z.object({
  id: z.string(),
  name: z.string(),
  kind: z.enum(["oidc", "saml"]),
  enabled: z.boolean(),
  customer_id: z.string().nullable().optional(),
  config: z.record(z.unknown()).default({}),
  created_at: z.string(),
  updated_at: z.string(),
});

export const AppIdpBindingSchema = z.object({
  id: z.string(),
  app_identity_provider_id: z.string(),
  app_identity_provider_name: z.string(),
  app_identity_provider_kind: z.string(),
  app_identity_provider_enabled: z.boolean(),
  customer_id: z.string(),
  application_id: z.string().nullable().optional(),
  created_at: z.string(),
  updated_at: z.string(),
});

export const LoginOptionsSchema = z.object({
  local_suffix: z.string(),
  suffixes: z.array(
    z.object({
      suffix: z.string(),
      auth_source_id: z.string(),
      auth_source_name: z.string(),
      kind: z.enum(["local", "oidc"]),
      sso: z.boolean(),
    }),
  ),
});

export type AuthSource = z.infer<typeof AuthSourceSchema>;
export type UpnSuffix = z.infer<typeof UpnSuffixSchema>;
export type AppIdentityProvider = z.infer<typeof AppIdentityProviderSchema>;
export type AppIdpBinding = z.infer<typeof AppIdpBindingSchema>;
export type LoginOptions = z.infer<typeof LoginOptionsSchema>;

export type AuthSourceCreatePayload = {
  name: string;
  kind: "oidc";
  enabled?: boolean;
  description?: string;
  issuer_url: string;
  client_id: string;
  client_secret?: string;
  scopes?: string;
  claim_username?: string;
  claim_groups?: string;
};

export type AuthSourceUpdatePayload = {
  name?: string;
  enabled?: boolean;
  description?: string;
  issuer_url?: string;
  client_id?: string;
  client_secret?: string;
  scopes?: string;
  claim_username?: string;
  claim_groups?: string;
};

File: backend/app/schemas/auth_sources.py
from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field

AuthSourceKind = Literal["local", "oidc"]
AppIdPKind = Literal["oidc", "saml"]


class AuthSourceRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    name: str
    kind: AuthSourceKind
    enabled: bool
    description: str
    issuer_url: str | None = None
    client_id: str | None = None
    has_client_secret: bool = False
    scopes: str
    claim_username: str
    claim_groups: str
    created_at: datetime
    updated_at: datetime


class AuthSourceCreate(BaseModel):
    name: str = Field(min_length=1, max_length=64)
    kind: AuthSourceKind = "oidc"
    enabled: bool = True
    description: str = Field(default="", max_length=512)
    issuer_url: str | None = Field(default=None, max_length=512)
    client_id: str | None = Field(default=None, max_length=256)
    client_secret: str | None = Field(default=None, max_length=512)
    scopes: str = Field(default="openid profile email", max_length=256)
    claim_username: str = Field(default="preferred_username", max_length=64)
    claim_groups: str = Field(default="groups", max_length=64)


class AuthSourceUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=64)
    enabled: bool | None = None
    description: str | None = Field(default=None, max_length=512)
    issuer_url: str | None = Field(default=None, max_length=512)
    client_id: str | None = Field(default=None, max_length=256)
    client_secret: str | None = Field(default=None, max_length=512)
    scopes: str | None = Field(default=None, max_length=256)
    claim_username: str | None = Field(default=None, max_length=64)
    claim_groups: str | None = Field(default=None, max_length=64)


class UpnSuffixRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    suffix: str
    auth_source_id: str
    auth_source_name: str = ""
    auth_source_kind: str = ""
    created_at: datetime
    updated_at: datetime


class UpnSuffixCreate(BaseModel):
    suffix: str = Field(min_length=1, max_length=255)
    auth_source_id: str


class UpnSuffixUpdate(BaseModel):
    auth_source_id: str | None = None


class AppIdentityProviderRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    name: str
    kind: AppIdPKind
    enabled: bool
    customer_id: str | None = None
    config: dict[str, Any] = Field(default_factory=dict)
    created_at: datetime
    updated_at: datetime


class AppIdentityProviderCreate(BaseModel):
    name: str = Field(min_length=1, max_length=128)
    kind: AppIdPKind = "oidc"
    enabled: bool = True
    customer_id: str | None = Field(default=None, max_length=64)
    config: dict[str, Any] = Field(default_factory=dict)


class AppIdentityProviderUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=128)
    kind: AppIdPKind | None = None
    enabled: bool | None = None
    customer_id: str | None = Field(default=None, max_length=64)
    config: dict[str, Any] | None = None


class LoginOptionsSuffix(BaseModel):
    suffix: str
    auth_source_id: str
    auth_source_name: str
    kind: AuthSourceKind
    sso: bool


class LoginOptionsResponse(BaseModel):
    local_suffix: str
    suffixes: list[LoginOptionsSuffix]
